src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -465,93 +426,165 @@ protected void InitializeReferences()
             }
         }
 
-        internal void Run<TState>(string input, int startat, ref TState state, MatchCallback<TState> callback, bool reuseMatchObject)
+        /// <summary>Internal worker which will scan the passed in string <paramref name="input"/> for all matches, and will call <paramref name="callback"/> for each match found.</summary>
+        internal void RunAllMatchesWithCallback<TState>(string input, int startat, ref TState state, MatchCallback<TState> callback, bool reuseMatchObject)
         {
             Debug.Assert((uint)startat <= (uint)input.Length);
 
             RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
             try
             {
-                runner.InitializeTimeout(internalMatchTimeout);
+                // For the string overload, we need to set runtext before starting the match attempts.
                 runner.runtext = input;
-                int runtextpos = startat;
-                while (true)
-                {
-                    runner.InitializeForScan(this, input, startat, false);
-                    runner.runtextpos = runtextpos;
+                RunAllMatchesWithCallbackHelper(input, startat, ref state, callback, runner, usingStringOverload: true, reuseMatchObject);
+            }
+            finally
+            {
+                runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache.
+                _runner = runner;
+            }
```

Comment:
Not for this PR, but subsequently I think we should delete the try/finally.  We never expect exceptions to come out of the execution, and if they do, the worst that happens is we lose the runner instance.